### DISCUSSION TITLE
awc::Client is not Send

### DISCUSSION THREAD

**Discussion by semtexzv** [actix/actix-web#2809] [Ideas]
After trying to use `awc` in an application with `tonic`, I've encountered an issue, where enything the client does is not `Send`. I've tracked it down to the fact that the `default_headers` [here](https://github.com/actix/actix-web/blob/50894e392ea5b6ad83bfc723139ff5a524913127/awc/src/client/mod.rs#L59) uses `Rc` instead of `Arc`.

I understand the choice to share these between different clients, but the inability to use the `Client` in `Send` contexts is really stopping me from using it. 

Why not use `Arc`? If there is only a single thread using the client, the overhead is minimal (Atomic operations in non-contended environments have low, non-zero overhead). And it adds the option to use client in Multi-threaded contexts (eg. My gRPC application performing http requests).

**Comment by robjtede** [comment]
Making the client Send is planned but for now I'd suggest using `reqwest`.

**Comment by fakeshadow** [comment]
awc is mostly meant to be used inside actix-web. it's short for actix-web-client. So why pay for thread safety when we don't need it.

That said making awc send `Send` is mostly about the client object being `Send`. The request and response futures it produces are likely to remain `!Send` to keep overhead low and compat with other actix-x crates. Which means it will have problem working with tonic still.

In general if you need `Send` futures you can use reqwest.
